tools/wake/config.cpp

```diff
@@ -0,0 +1,313 @@
+      continue;
+    }
+    if (size < 0) {
+      perror("failed to exec /bin/sh");
+      exit(EXIT_FAILURE);
+    }
+    if (size == 0) {
+      break;
+    }
+    acc.append(buf, buf + size);
+  }
+
+  // We should be done already but let's wait for the child and
+  // make sure everything worked well there.
+  int status = 1;
+  if (waitpid(pid, &status, 0) != pid) {
+    perror("waitpid: failed to wait /bin/sh -c echo ...");
+    exit(EXIT_FAILURE);
+  }
+  if (status != 0) {
+    fprintf(stderr, "waitpid: /bin/sh -c echo ... failed with non-zero exit status");
+    exit(EXIT_FAILURE);
+  }
+
+  // Remove the newline added by echo
+  // echo -n doesn't work MacOS
+  acc.pop_back();
+
+  // Return the string we got back
+  return acc;
+}
+
+// Find the default location for the user level wake config
+static std::string default_user_config() {
+  // If XDG_CONFIG_HOME is set use it, otherwise use home
+  char* xdg_config_home = getenv("XDG_CONFIG_HOME");
+  char* home_dir = getenv("HOME");
+
+  if (home_dir == nullptr) {
+    std::cerr << "$HOME is not set!" << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  std::string prefix = wcl::join_paths(std::string(home_dir), ".config");
+
+  if (xdg_config_home != nullptr) {
+    prefix = std::string(xdg_config_home);
+  }
+
+  return wcl::join_paths(prefix, ".wake.json");
+}
+
+enum class ReadJsonFileError {
+  BadFile,
+  InvalidJson,
+};
+
+static wcl::result<JAST, std::pair<ReadJsonFileError, std::string>> read_json_file(
+    const std::string& path) {
+  std::ifstream file(path);
+  if (!file) {
+    return wcl::make_error<JAST, std::pair<ReadJsonFileError, std::string>>(
+        ReadJsonFileError::BadFile, "Failed  to read '" + path + "'");
+  }
+
+  std::stringstream buff;
+  buff << file.rdbuf();
+  std::string contents = buff.str();
+
+  JAST json;
+  std::stringstream errors;
+  if (!JAST::parse(contents, errors, json)) {
+    return wcl::make_error<JAST, std::pair<ReadJsonFileError, std::string>>(
+        ReadJsonFileError::InvalidJson, path + " must be a valid JSON object: " + errors.str());
+  }
+
+  return wcl::result_value<std::pair<ReadJsonFileError, std::string>>(std::move(json));
+}
+
+static std::vector<std::string> find_disallowed_keys(const JAST& json,
+                                                     const std::set<std::string>& keys) {
+  std::vector<std::string> disallowed = {};
+
+  for (const auto& key : keys) {
+    if (json.get(key).kind == JSON_NULLVAL) {
+      continue;
+    }
+    disallowed.push_back(key);
+  }
+
+  return disallowed;
+}
+
+bool init(const std::string& wakeroot_path) {
+  if (_config != nullptr) {
+    std::cerr << "Cannot initialize config twice" << std::endl;
+    assert(false);
+  }
+
+  // Keys that may not be specified in .wakeroot
+  std::set<std::string> wakeroot_disallowed_keys = {};
+
+  // Keys that may not be specified in the user config
+  std::set<std::string> user_config_disallowed_keys = {"version", "user_config"};
```

Comment:
I think we should also have a set of allowed keys so that if someone adds a wrong key they get at least a warning if not an error.

Reply:
Yeah some people move backward/forward on wake versions depending on what they're working on, would be nice if they could use the same homedir config without it erroring out

Reply:
I'm happy with that.